packages/engine/core/src/lib/abstract.ts

```diff
@@ -77,48 +77,39 @@ export class Plugin<T extends Api = any, App extends ApiMap = any> implements Pl
     return this[method](...args)
   }
 
+  protected setCurrentRequest(request: PluginRequest) {
+    this.currentRequest = request
+  }
+
+  protected letContinue() {
+    delete this.currentRequest
+    this.queue = this.queue.filter((value) => {
+      return value.canceled === false && value.timedout === false && value.finished === false
+    })
+    const next = this.queue.find((value) => {
+      return value.canceled === false && value.timedout === false && value.finished === false
+    })
+    if (next) next.run()
+  }
+
   /** Add a request to the list of current requests */
   protected addRequest(request: PluginRequest, method: Profile<T>['methods'][number], args: any[]) {
     return new Promise((resolve, reject) => {
-      // Add a new request to the queue
-      this.requestQueue.push(async () => {
-        this.currentRequest = request
-        let timedout = false
-        const letcontinue = () => {
-          if (timedout) {
-            const { from } = this.currentRequest
-            const params = args.map(arg => JSON.stringify(arg)).join(', ')
-            const error = `[TIMED OUT]: Call to method "${method}" from "${from}" to plugin "${this.profile.name}" has timed out with arguments ${params}."`
-            reject(error)
-          }
-          // Remove current request and call next
-          delete this.currentRequest
-          this.requestQueue.shift()
-          if (this.requestQueue.length !== 0) this.requestQueue[0]()
-        }
-
-        const ref = setTimeout(() => {
-          timedout = true
-          letcontinue()
-        }, this.options.queueTimeout || 10000)
-
-        try {
-          const result = await this.callPluginMethod(method, args)
-          delete this.currentRequest
-          if (timedout) return
-          resolve(result)
-        } catch (err) {
-          delete this.currentRequest
-          reject(err)
-        }
-        clearTimeout(ref)
-        letcontinue()
-      })
-      // If there is only one request waiting, call it
-      if (this.requestQueue.length === 1) {
-        this.requestQueue[0]()
-      }
-    })
+      const queue = new PluginQueue(resolve, reject, request, method, this.options, args)
```

Comment:
maybe `PluginQueueItem` is preferable here (first I thought that this object contains the list of pending requests)